sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -190,6 +194,14 @@ RetT *getOrBuild(KernelProgramCache &KPCache, const KeyT &CacheKey,
     KPCache.notifyAllBuild();
 
     return Desired;
+  } catch (const exception &Ex) {
+    using BuildResultT = KernelProgramCache::BuildResultT;
+    WithState->BuildResult.reset(new BuildResultT{Ex.what(), Ex.get_cl_code()});
+    WithState->State.store(BS_Failed);
+
+    KPCache.notifyAllBuild();
+
+    std::rethrow_exception(std::current_exception());
   } catch (...) {
     WithState->State.store(BS_Failed);
```

Comment:
Not sure what happens if we reach here: the program/kernel is null, but BuildResult is not set. I suspect that the assert on line 139 fires in this case.
> ```
> if (WithBuildState->BuildResult.get()) {
>   ...
> }
> RetT *Result = WithBuildState->Ptr.load();
> assert(Result && "An exception should have been thrown");
> ```

Reply:
Yes, this assert will be raised upon second attempt to get kernel/program. Though if any exception is thrown from build process we still have to set state to BS_Failed. Awaiting process won't end within the users' lifetime otherwise.